src/service/ampdoc-impl.js

```diff
@@ -0,0 +1,219 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../log';
+import {getService} from '../service';
+import {isShadowRoot} from '../types';
+
+/** @const {string} */
+const AMPDOC_PROP = '__AMPDOC';
+
+
+/**
+ * This service helps locate an ampdoc (`AmpDoc` instance) for any node,
+ * either in the single-doc or shadow-doc environments.
+ *
+ * In the single-doc environment an ampdoc is equivalent to the
+ * `window.document`. In the shadow-doc mode, any number of AMP documents
+ * could be hosted in shadow roots in the same global `window.document`.
+ *
+ * @package
+ */
+export class AmpDocService {
+  /**
+   * @param {!Window} win
+   * @param {boolean} isSingleDoc
+   */
+  constructor(win, isSingleDoc) {
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @private @const {?AmpDoc} */
+    this.singleDoc_ = null;
+    if (isSingleDoc) {
+      this.singleDoc_ = new AmpDocSingle(ths.win);
+    }
+  }
+
+  /**
+   * Whether the runtime in the single-doc mode. Alternative is the shadow-doc
+   * mode that supports multiple documents per a single window.
+   * @return {boolean}
+   */
+  isSingleDoc() {
+    return !!this.singleDoc_;
+  }
+
+  /**
+   * Returns the instance of the ampdoc (`AmpDoc`) that contains the specified
+   * node. If the runtime is in the single-doc mode, the one global `AmpDoc`
+   * instance is always returned. Otherwise, this method locates the `AmpDoc`
+   * that contains the specified node and, if necessary, initializes it.
+   *
+   * @param {!Node} node
+   * @return {!AmpDoc}
+   */
+  getAmpDoc(node) {
+    // Single document: return it immediately.
+    if (this.singleDoc_) {
+      return this.singleDoc_;
+    }
+
+    // A custom element may already have the reference to the ampdoc.
+    if (typeof node.getAmpDoc == 'function') {
+      const ampDoc = node.getAmpDoc();
+      if (ampDoc) {
+        return ampDoc;
+      }
+    }
+
+    // Otherwise discover and possibly create the ampdoc.
+    const shadowRoot = findShadowRoot(node);
+    if (!shadowRoot) {
+      throw dev.createError('No root found for', node);
+    }
+    let ampDoc = shadowRoot[AMPDOC_PROP];
+    if (!ampDoc) {
+      ampDoc = new AmpDocShadow(shadowRoot);
+      shadowRoot[AMPDOC_PROP] = ampDoc;
+    }
+    return ampDoc;
+  }
+}
+
+/**
+ * @param {!Node} node
+ * @return {?ShadowRoot}
+ */
+function findShadowRoot(node) {
+  // TODO(dvoytenko): Replace with `getRootNode()` API when it's available.
+  for (let n = node; n; n = n.parentNode) {
```

Comment:
Not sure this needs to be optimized? In that case, is there any chance we could use closest?